iocore/net/quic/QUICStreamState.h

```diff
@@ -65,94 +62,93 @@ class QUICStreamState
   virtual void update_with_sending_frame(const QUICFrame &frame)   = 0;
   virtual void update_with_receiving_frame(const QUICFrame &frame) = 0;
 
+  virtual void
+  update_on_transport_send_event()
+  {
+  }
+  virtual void
+  update_on_transport_recv_event()
+  {
+  }
+  virtual void
+  update_on_user_read_event()
+  {
+  }
+
   virtual bool is_allowed_to_send(QUICFrameType type) const        = 0;
   virtual bool is_allowed_to_send(const QUICFrame &frame) const    = 0;
   virtual bool is_allowed_to_receive(QUICFrameType type) const     = 0;
   virtual bool is_allowed_to_receive(const QUICFrame &frame) const = 0;
 
 protected:
-  void _set_state(State s);
-
-private:
-  State _state = State::_Init;
-};
-
-class QUICUnidirectionalStreamState : public QUICStreamState
-{
-public:
-  QUICUnidirectionalStreamState(QUICTransferProgressProvider *in, QUICTransferProgressProvider *out)
-    : _in_progress(in), _out_progress(out)
+  void
+  _set_state(T s)
   {
+    ink_assert(s != T::Init);
+    this->_state = s;
   }
-  virtual void update(const QUICStreamState &opposite_side) = 0;
 
-protected:
-  QUICTransferProgressProvider *_in_progress  = nullptr;
```

Comment:
What is your motivation of removing `QUICTransferProgressProvider` ?

Reply:
Because we can use the `update_on_transport_recv_event`, `update_on_transport_send_event` and `update_on_user_read_event` instead. 

And `QUICTransferProgressProvider` has strongly dependencies with VIO. But we can avoid it.

See `QUICIncomingStreamFrameBuffer` and `QUICStream::on_read` 
```
void
QUICStream::on_read()
{
  if (this->_received_stream_frame_buffer.empty() && this->_received_stream_frame_buffer.has_all_data()) {
    this->_state.update_on_user_read_event();
  }
}
```

That means all data has been read by user.

Reply:
That's not true. `QUICTransferProgressProviderVIO` has dependency for VIO but `QUICTransferProgressProvider` doesn't. I introduced `QUICTransferProgressProvider` interface to hide the underlying implementation and to provide one common interface. You don't need to introduce `has_all_data`.

Reply:
Okay, I bring it back.